### DISCUSSION TITLE
Server-only paths config

### DISCUSSION THREAD

**Discussion by nakleiderer** [remix-run/remix#3451] [Proposals]
Remix's automatic server vs. client code-splitting is amazing. However, when learning Remix it can be a bit of a foot gun. New and experienced devs might end up accidentally shipping server-only code to the client. Depending on the code, it might not cause an error in development and may potentially leak information about business logic, database structures, or other information that's best kept secret.

The `.server.ts` file extension is a great way to prevent code getting to the client, but it's a bit verbose to sprinkle everywhere. It would be convenient to be able to mark server-only modules in bulk. 

**Proposals:**
Here's a few ways Remix could help:
- A folder syntax for marking files as server-only. Maybe any folder ending with `.server` will signal that all files contained inside should be ignored. ie. `/app/models.server/user.ts`
- A config option in `remix.config.js`. Maybe a glob of files could be marked as server only. ie: `{ serverOnlyPaths: ["app/models/**", "app/sensitive-stuff/**"] }`

**Possible Workarounds:**
- Put an `index.server.ts` file in the folder intended for server-only use and re-export any subfiles. Make sure to only use the `index.server` in imports. This has similar problems because the sub-modules can still be directly imported, but maybe a linting rule could be made to enforce with for a team. 

Thanks to @kiliman for helping me think this through in Discord.

**Comment by kiliman** [comment]
I created a patch that adds support for server *folders*. Just append `.server` to your folder name and all the files in it will be treated as server-only.

https://github.com/kiliman/remix-server-folders
